posit-dev/positron
# Re-use Data Explorer windows instead of opening new ones

### Intent

When the user requests to view a variable that's already open in a data explorer, activate the existing data explorer instead of opening a new one.

Addresses https://github.com/posit-dev/positron/issues/2303

### Approach

When the `view` message opens a comm, the runtimes now return the ID of the new comm (the Python change is included in the PR; the R change is in https://github.com/posit-dev/amalthea/pull/350). This allows us to generate a binding between the variable in the Variables pane and the Data Explorer (which uses the comm as a transport). 

When the user views a variable in the Variables pane, we check the bindings to see if any data explorers are open for the variable before opening a new one. If we find any, we activate the existing one instead of sending the `view` RPC to open a new one.

As part of this change, I added disposal of the underlying comm when the tab closes. 

### QA Notes

- What I've implemented here is what @jthomasmock describes in this comment: https://github.com/posit-dev/positron/issues/2303#issuecomment-1980874747
- Only data sets opened from the Variables pane can be re-activated, and only opening a dataset from the Variables pane will re-activate an existing tab. Using `View()` (R) or `%view` (Python) will always open a new tab.
- If the data viewer is disconnected, then it shouldn't ever be re-activated (we have no reconnection behavior today). Instead, a new viewer should be opened.
- Every runtime instance should have a distinct set of tabs. That is, clicking on a variable named "foo" in R should not activate a data explorer bound to a variable named "foo" in Python. It is possible to have multiple data explorers with the same label open at the same time.
- Popping the data viewer into a new window makes things a little tricky, because the editor Kind of Closes but Not Really. Make sure that works okay.
